autonomy_core/map_plan/mapper/src/test_voxel_mapper.cpp

```diff
@@ -0,0 +1,428 @@
+    }
+    time_end = std::chrono::steady_clock::now();
+    std::cout << "CALLBACK Duration (decayLocalCloud): " <<
+        std::chrono::duration_cast<std::chrono::microseconds>(time_end - time_start).count()
+        << "us" << std::endl;
+
+    planning_ros_msgs::VoxelMap decayed_map = p_test_mapper_->getMap();
+
+    // Create the ground truth voxel map to compare it to the decayed map
+    planning_ros_msgs::VoxelMap gt_voxel_map;
+    int dim_x = x_dim_ / resolution_;
+    int dim_y = y_dim_ / resolution_;
+    int dim_z = z_dim_ / resolution_;
+    // All voxels should be marked as occupied except the decayed voxel locations
+    gt_voxel_map.data.resize(dim_x * dim_y * dim_z, p_test_mapper_->val_occ);
+
+    // Free the voxels that should be decayed
+    for (int x = 180; x < 219; x++)
+    {
+        for (int y = 180; y < 219; y++)
+        {
+            for (int z = 0; z < 20; z++)
+            {
+                int idx = x + 400 * y + 400 * 400 * z;
+                gt_voxel_map.data[idx] = p_test_mapper_->val_free;
+            }
+        }
+    }
+
+    // Finally make the comparison
+    ASSERT_EQ(decayed_map.data.size(), gt_voxel_map.data.size());
+    for (int idx = 0; idx < num_voxels; idx++)
+    {
+        EXPECT_EQ(decayed_map.data[idx], gt_voxel_map.data[idx]) << "IDX: " << idx;
+    }
+}
+
+/**
+ * @brief This method is called to process the point cloud from the lidar scans. Currently,
+ * raytracing is always set to false, so it is never called. The decayLocalCloud method is called
+ * inside this method, but since the decay value is 2, it does not impede a voxel from being marked
+ * as occupied after one scan if we include 3 lidar points in the specified voxel. Points are
+ * provided with respect to the lidar frame and the transformation that is passed as an argument
+ * is the pose of the lidar in the map frame. Any points outside of max_range are discarded.
+ * In this test, we simulate the lidar to be at a world coordinate different than (0, 0, 0) and 
+ * make a single call to addCloud. This represents a single lidar scan, but to make sure that
+ * the voxels corresponding to the points are marked as occupied in this single scan, we triplicate
+ * the points. After making the call we compare the normal map and the inflated map to their ground
+ * truths.
+ */
+TEST_F(VoxelMapperTest, TestAddCloud)
+{
+    // We will consider that the lidar is at position (-10, -5, -1) in the map frame of reference
+    Eigen::Vector3d lidar_pos(-10, -5, -1);
+    Eigen::Affine3d t_map_lidar = Eigen::Translation3d(-10, -5, -1) *
+                                  Eigen::AngleAxisd(0, Eigen::Vector3d(0, 0, 0));
+    double max_range = 30.0;
+    vec_Vec3d lidar_points;
+
+    // Add 3 points for all voxels in the range -30 to in all 3 axes. Some will be discarded
+    // and only a sperical shape of voxels around (-10, -5, -1) should be marked as occupied
+    for (double x = -30; x <= 30; x+=0.5)
```

Comment:
Recommend only using integer type in for loop

Reply:
I thought of two other ways to do it because I can see how a double in a for loop looks uncommon, but I found this way was more code readable. Here are the other two ways that I thought of, please let me know if you think I should replace the above code with one of these. First method seems less readable and second method is just replacing for loop with while loop. 
Method 1:
```
for (int x = 0; x <=120; x++)
{
    double real_x = -30 + x * 0.5;
    // use real_x for computations
}
```

Method 2:
```
double x = -30;
while(x <=30)
{
    // use x for computations
    x += 0.5;
}
```

Reply:
1st one is the right one. 
Never put floating-point comparison in loop statement.